#163586540 Create party

frankly034 opened the pull request:
#### What does this PR do?
Have api endpoint for POST /parties

#### Description of Task to be completed?
* Create party model data
* Create party model
* Create party controller
* Create party middleware 
* Create party route
* Create party create test

#### How should this be manually tested?
After cloning the repo, run 'npm run test'

#### Any background context you want to provide?
* Http test were done using supertest and chai expect library

#### What are the relevant pivotal tracker stories?
#163586540

frankly034 committed:
chore(party data): adds dummy party data

- adds an array of parties
[Delivers #163586540]

frankly034 committed:
chore(party data): adds dummy party data

- adds an array of parties
[Delivers #163586540]

frankly034 committed:
chore(partyModel): implement partyModel create()

- adds a valid party to the dummy data
- returns the inserted party

frankly034 committed:
chore(partyModel): implement partyModel create()

- adds a valid party to the dummy data
- returns the inserted party

frankly034 committed:
bug(partyID) Adds ID for new party

- ensures that the newly created party has an incremented ID
[Delivers #163586540]

frankly034 committed:
bug(partyID) Adds ID for new party

- ensures that the newly created party has an incremented ID
[Delivers #163586540]

frankly034 committed:
chore(partyController): create createParty()

- checks for req.error and attaches the appropriate error code and
message
- creates a new party with status 200
- returns the created party object

frankly034 committed:
chore(partyController): create createParty()

- checks for req.error and attaches the appropriate error code and
message
- creates a new party with status 200
- returns the created party object

frankly034 committed:
chore(partyMiddleware) added create() middleware

- ensure that all required inputs are passed from req.body
- attaches appropriate req.err object otherwise

[Delivers #163586540]

frankly034 committed:
chore(partyMiddleware) added create() middleware

- ensure that all required inputs are passed from req.body
- attaches appropriate req.err object otherwise

[Delivers #163586540]

frankly034 committed:
chore(partyRoutes) created POST /parties

- setup the route file
- created the POST /parties routes with middleware and controller

frankly034 committed:
chore(partyRoutes) created POST /parties

- setup the route file
- created the POST /parties routes with middleware and controller

frankly034 committed:
feat(POST /parties) implement create parties api

- adds partyRoutes to server.js
- created express middleware for the route
[Delivers #163586540]

frankly034 committed:
chore(test script): modify test start script

- set test start script with mocha, babel and nodemon
[Delivers #163586540]

frankly034 committed:
chore(app): export app

- exports app so it can be reused
[Delivers #163586540]

frankly034 committed:
feat(test): setup post /parties test

- ensure that all required valid attributes are passed
[Delivers #163586540]

frankly034 committed:
feat(test): setup post /parties test

- ensure that all required valid attributes are passed
[Delivers #163586540]

Lumexralph commented on server/controllers/partyController.js:
```diff
@@ -0,0 +1,17 @@
+  static createParty(req, res) {
+    if (req.error) {
+      return res.status(req.error.status).send({ status: req.error.status, error: req.error.msg });
+    }
```
@frankly034 Nice work! You can further improve on your work by storing `req.error` in a variable so as to avoid a lookup every time it is used e.g `const error = req.error;` and then use `error`, what do you think about this?

frankly034 replied:
@Lumexralph Thanks. 
I see what you said. Your observation is in line. 
I've implemented the change.

Lumexralph commented on server/routes/partyRoutes.js:
```diff
@@ -0,0 +1,11 @@
+
+const router = express.Router();
+
+const apiStr = '/api/v1/parties';
```
@frankly034 let the naming of this reflect what value it's holding, what do you think about doing `const apiUrl` and also ensure to version your API.

frankly034 replied:
@Lumexralph Thanks for the observation, was actually looking for a very descriptive name. I've made the changes and pushed

frankly034 committed:
bug(partyController): Refactored code

- remove repeated call to req.error
[Delivered #163586540]

frankly034 committed:
bug(partyController): Refactored code

- remove repeated call to req.error
[Delivered #163586540]

frankly034 committed:
bug(partyRoutes): descriptive var naming

- Proper descriptive naming for URL version
[Delivers #163586540]

Lumexralph commented on server/routes/partyRoutes.js:
```diff
@@ -0,0 +1,11 @@
+const router = express.Router();
+
+const apiUrl = '/api/v1/parties';
+
```
@frankly034 Nice work can please `version` your API?

frankly034 replied:
I've versioned the api. Love to have your review on it. Thanks

frankly034 committed:
chore(versioning): version api route

- implemented api versioning
[Delivers #163586540]

frankly034 committed:
chore(officeData): add office dummy data

- adds a dummy js data structure to represent data
[Delivers #163600428]

frankly034 committed:
chore(officeModel): create office model

- create() officeModel to create new office data

frankly034 committed:
chore(officeController): create createOffice()

- add createOffice() controller function
- handle create office

frankly034 committed:
chore(middleware): office middleware

- create office middleware
- implemented the create() for validating required middleware attributes

frankly034 committed:
feat(server): implement POST /offices

- setup the server to listen for a post request to /offices
- add a new government office entry to dummy data

frankly034 committed:
bug(model): Fix a typo

- fix the passed in object typo
- hqAddress to type

frankly034 committed:
chore(route): create office route function

- add office route
- create the post /office route
- exports router
[Delivers #90304342]

frankly034 committed:
chore(test): add POST /offices

- tests if all the required fields are entered
[Delivers #163600428]

frankly034 committed:
bug(test): modified test start script

- Removed --reporter as it is not yet installed
[Delivers #1636004283]

Lumexralph committed:
Merge pull request #21 from frankly034/ft-create-office-endpoint-163600428

#163600428 Create office endpoint

frankly034 committed:
bug(rebase): resolved conflicts

- resolve merge conflicts
- refactor middlewares to return error response instead of sending to
controller
- versioned the party route
[Delivers #163586540]

frankly034 committed:
bug(merge): resolve merge conflict

Merge branch 'ft-create-party-endpoint-163586540' of https://github.com/frankly034/politico into ft-create-party-endpoint-163586540
[Delivers #163586540]

frankly034 commented:
Fixed the middleware return cases for both the partyMiddleware and the officeMiddleware, made sure the versioning was still in place.

Used the rebase like you told me and resolved the merge conflicts. Pulled, merged and resolved conflicts before pushing the updated version. 

Please review the changes.

Thanks.

Lumexralph reviewed (approved):
Good Job! This work looks good to me.
